### ISSUE TITLE
Upgrade warning message is too intrusive; magit-last-seen-setup-instructions

### ISSUE THREAD

**Issue by benma** [magit/magit#1803]
Hi

I just upgraded magit on melpa and am greeted with this message on emacs startup:

```You have just updated to version 1.4.0 of Magit, and have to
make a choice.```

To make that disappear, it instructs me to add this to my config:

```(setq magit-last-seen-setup-instructions "1.4.0")```

This behavior is too intrusive and annoying. The package and this message are not so important to force everyone to deal with this. Making people aware of `(setq magit-auto-revert-mode nil)` is not important enough to annoy people with this manual intervention, especially since it has been the default behavior before.

I strongly dislike this move. Imagine if every package started the habit of forcing the user to modify their config files to get rid of intrusive messages. If it is a bad idea for other packages, it is a bad idea for this package.

I suggest just dropping this message and config variable.

At the very least, please display the message only once and never again, without forcing the user to suppress the message manually.

**Comment by tarsius** [comment]
Okay now we have this situation:

1. I, the maintainer of the package, think that it is important that this feature is enabled, because otherwise many things would not work as intended me and also because by now it is expected by most users.

2. One user comes along and argues that it is a very very dangerous feature, and that that enabling it by default is very bad. See #1783.

   I think this is only partially true, because (a) the action effect is reversible and (b) its effect is only undesirable when using a workflow which, in my opinion, doesn't make much sense when using Magit. 

   Never-the-less I do agree that this can take a user by surprise, especially because the effect of the feature can go unnoticed until it is to late to undo it. Also I am in the final steps of creating the first release in a very long time. I chose to inform users about the issue. This way most users will continue to use Magit as intended while those who would be harmed by it, are given the opportunity to turn it off, before anything bad happens.

   While I think this is an important feature, it also isn't absolutely essential, and if it conflicts with the way one uses Emacs, then it is perfectly fine to disable it. And I don't want to responsible for dataloss. (You got annoyed because I am forcing you to set a variable - imagine how annoyed someone would get because I am responsible for them losing valuable work).

3. Another user comes along claiming it is wrong of me to force users to make a choice.

   The way I see it, I have no other choice. Turning the feature of by default is a no-go because doing so would break many workflows. Also many users don't read release notes and faqs. So turning it of would cause many bug reports which I then would have to spend my time on.

   At the same time it is dangerous for users who are completely unaware of its existence and I simply don't want to be accused of being responsible for them losing work. This feature existed before I became Magit's maintainer, but this is the first release I am making. `1.4.0` isn't a development branch, it is a release; and as the one who created it, I am the one who will be blamed if it does damage.

Having to set a variable is a very very small inconvenience. You should be willing to make that sacrifice. Not forcing you to do it, could mean that other users would lose hours worth of work.

> The package and this message are not so important to force everyone to deal with this.

The message is very important - it makes sure users are aware of the issue. The package also is important - it's the most popular (end-user) package on Melpa, by a huge margin.

> If it is a bad idea for other packages, it is a bad idea for this package.

I disagree with both parts of that sentence. If a package does something dangerous I want to be informed about that. And that is especially true for a popular package like Magit.

> I suggest just dropping this message **and config variable**.

You get annoyed by having to set an option. Imagine how you would feel if I forced you to change your workflow.

> At the very least,

Never say that. These are words that very quickly and reliably piss of a maintainer who has invested countless hours maintaining a tool you are using for free.

> please display the message only once and never again, without forcing the user to suppress the message manually.

That's a reasonable suggestion. Maybe I will do that, maybe not.

**Comment by npostavs** [comment]
> this message on emacs startup

I was going to say, it should be delayed until magit loads, but it seems to be implemented with an [`after-init-hook`][hook] which *is* being set at load time. Wouldn't this fail to trigger if magit isn't loaded during init? And isn't magit only autoloaded during init by default? 

#### So most users won't see the message?

---

> without forcing the user to suppress the message manually.

@tarsius: didn't you have some similar sort of thing with the popup config (something with the prefix arg), where you used the customize machinery? Then users too lazy to edit their init files could just click a button.

[hook]: https://github.com/magit/magit/blob/1d0d054c04a168a3ba39bf77c54d06ba1d025d2e/magit.el#L7818

**Comment by tarsius** [comment]
@npostavs Unfortunately you are right about this, and I consider it a serious issue which has to be fixed quickly. I have done so and also have created the bugfix release `1.4.1`. I would have liked to spend some more time on it, but since I am leaving for a few days tomorrow, I had to do it now.

Yes, I could do something similar like for for the popup option, but there wasn't enough time to do it now (I should get some sleep now). ... there likely will be a `1.4.2`.

Thanks for pointing these things out.

**Comment by benma** [comment]
@tarsius 

First of all, thanks for taking the time to write down this detailed response. I wrote the issue in a hurry, I apologize that it came across so harsh. 

> I disagree with both parts of that sentence. If a package does something dangerous I want to be informed about that. And that is especially true for a popular package like Magit.

* What about making the non-dangerous setting the default, and recommend the dangerous setting in the installation instructions? 
* What about asking the user with a y-or-n question the first time that setting is used, instead of at startup? Similar in how multiple-cursors asks how to use an operation it does not know about yet.

Anyway, I agree it is no big deal. It just seemed wrong to make every user go through that decision and manual work when most have used the package before and know how it works.

Thanks for all the work you put into this package, I really do appreciate it! 

@tarsius  $3 @changetip

**Comment by changetip** [comment]
Hi @tarsius, @benma sent you a Bitcoin tip worth 12,737 bits ($3.00), and I'm here to deliver it ➔ **[collect your tip](https://www.changetip.com/collect/706644)**.


**[Learn more about ChangeTip](https://www.changetip.com/tip-online/github)**

**Comment by tarsius** [comment]
@benma Sorry I overreacted. And thanks for the tip.

I have opened a new issue #1809 to discus whether the mode should be on or off, and if the user should be asked what she wants.

Lets us *this* issue only to discus, *how* to ask the user *provided* we decide we should do that at all.

**Comment by tarsius** [comment]
I will reopen if we/I decide to go down that road.
